src/libraries/System.Net.Http/src/System/Net/Http/Headers/ObjectCollection.cs

```diff
@@ -10,41 +10,33 @@ namespace System.Net.Http.Headers
     /// <summary>An <see cref="ICollection{T}"/> list that prohibits null elements and that is optimized for a small number of elements.</summary>
     [DebuggerDisplay("Count = {Count}")]
     [DebuggerTypeProxy(typeof(ObjectCollection<>.DebugView))]
-    internal sealed class ObjectCollection<T> : ICollection<T> where T : class
+    internal class ObjectCollection<T> : ICollection<T> where T : class
     {
         private const int DefaultSize = 4;
 
-        /// <summary>Optional delegate used to validate added items.</summary>
-        private readonly Action<T>? _validator;
         /// <summary>null, a T, or a T[].</summary>
         internal object? _items;
         /// <summary>Number of elements stored in the collection.</summary>
         internal int _size;
 
         public ObjectCollection() { }
 
-        public ObjectCollection(Action<T> validator) => _validator = validator;
-
         public int Count => _size;
 
         public bool IsReadOnly => false;
 
-        public void Add(T item)
+        public virtual void Validate(T item)
         {
-            // Validate the item, either just by checking it for null, or using a custom validator,
-            // which should also check for null.
-            if (_validator is null)
+            if (item is null)
             {
-                if (item is null)
-                {
-                    throw new ArgumentNullException(nameof(item));
-                }
-            }
-            else
-            {
-                _validator.Invoke(item);
-                Debug.Assert(item != null);
+                throw new ArgumentNullException(nameof(item));
             }
+        }
+
+        public void Add(T item)
+        {
+            Validate(item);
```

Comment:
The combination of unsealing the type and adding this Validate is now introducing virtual calls onto a path that didn't previously have them.  The purpose of this PR is to improve perf, but this could regress it.  Can you share perf numbers for a variety of uses where the base and derived collection types are being used?

Reply:
There is a slight improvement, but it seems to be in a noise range:

|                             Method |        Job |                                                                                          Toolchain |      Mean |     Error |    StdDev | Ratio | RatioSD |  Gen 0 |  Gen 1 | Allocated |
|----------------------------------- |----------- |--------------------------------------------------------------------------------------------------- |----------:|----------:|----------:|------:|--------:|-------:|-------:|----------:|
|    Add_CallVirtValidation_FromBase | Job-HDJGGI |  /runtime_main/artifacts/bin/testhost/net7.0-OSX-Release-x64/shared/Microsoft.NETCore.App/7.0.0/corerun | 149.61 ns | 14.268 ns | 20.913 ns |  1.00 |    0.00 | 0.0050 | 0.0017 |      32 B |
|    Add_CallVirtValidation_FromBase | Job-FSEJVI | /runtime_pr/artifacts/bin/testhost/net7.0-OSX-Release-x64/shared/Microsoft.NETCore.App/7.0.0/corerun | 132.10 ns | 13.465 ns | 19.311 ns |  0.89 |    0.14 | 0.0050 | 0.0017 |      32 B |
|                                    |            |                                                                                                    |           |           |           |       |         |        |        |           |
| Add_CallVirtValidation_FromDerived | Job-HDJGGI |  /runtime_main/artifacts/bin/testhost/net7.0-OSX-Release-x64/shared/Microsoft.NETCore.App/7.0.0/corerun |  55.02 ns |  4.579 ns |  6.567 ns |  1.00 |    0.00 |      - |      - |         - |
| Add_CallVirtValidation_FromDerived | Job-FSEJVI | /runtime_pr/artifacts/bin/testhost/net7.0-OSX-Release-x64/shared/Microsoft.NETCore.App/7.0.0/corerun |  52.46 ns |  4.375 ns |  5.988 ns |  0.97 |    0.12 |      - |      - |         - |

```c#
using System.IO;
using System.Net.Http.Headers;
using BenchmarkDotNet;
using BenchmarkDotNet.Attributes;
using BenchmarkDotNet.Running;

[MemoryDiagnoser]
public class CacheControlHeaderValueBenchmarks
{
    public static void Main(string[] args) =>
        BenchmarkSwitcher.FromTypes(new []{ typeof(CacheControlHeaderValueBenchmarks) }).Run(args);

    private readonly CacheControlHeaderValue _cacheControl = new CacheControlHeaderValue();

    private readonly string _randomString = Path.GetRandomFileName();

    [Benchmark]
    public void Add_CallVirtValidation_FromBase() // CacheControlHeaderValue -> ObjectCollection -> ObjectCollection.Validate()
    {
        _cacheControl.Extensions.Add(new NameValueHeaderValue(_randomString));
    }

    [Benchmark]
    public void Add_CallVirtValidation_FromDerived() // CacheControlHeaderValue -> TokenObjectCollection -> ObjectCollection -> TokenObjectCollection.Validate()
    {
        _cacheControl.NoCacheHeaders.Add(_randomString);
    }
}
```

```sh
$ dotnet run -c Release -- --job ShortRun --iterationTime 100 --launchCount 2 --warmupCount 1 --filter '*' --corerun \
    ../runtime_main/artifacts/bin/testhost/net7.0-OSX-Release-x64/shared/Microsoft.NETCore.App/7.0.0/corerun \
    ../runtime_pr/artifacts/bin/testhost/net7.0-OSX-Release-x64/shared/Microsoft.NETCore.App/7.0.0/corerun
```

Reply:
There's a few places that use `ObjectCollection<NameValueHeaderValue>` with no validation and that are relatively common cases (`MediaTypeHeaderValue` for example). I wonder if having a `NameValueHeaderValueCollection` that's a sealed derived version of `ObjectCollection<NameValueHeaderValue>` would help as it might be able to do guarded devirt and inline the null check.